### PR TITLE
Spl2 fixes

### PR DESCRIPTION
**Pull request by fantavlik**
- Provide better error handling and configurability for language server download and install by allowing user to specify custom path to LSP file
- Fix issue with progress bars getting stuck
- Allow users to input earliest and latest times
Demo:
spl2-earliest-latest

- Serialize the format for SPL2 notebooks saved to disk to more closely match modules.json expectations, example:
```json
{
  "modules": [
    {
      "name": "module1",
      "namespace": "",
      "definition": "$q1 = from _internal where log_level != \"ERROR\"",
      "_vscode": {
        "metadata": {
          "splunk": {
            "earliestTime": "-24h",
            "latestTime": "now"
          }
        },
        "outputs": []
      }
    },
    {
      "name": "module2",
      "namespace": "",
      "definition": " $q2 = from main",
      "_vscode": {
        "metadata": {
          "splunk": {
            "earliestTime": "@d",
            "latestTime": "now"
          }
        },
        "outputs": []
      }
    }
  ],
  "app": "apps.search"
}
```

### CODE REVIEW

File: out/notebooks/commands.ts [hunk 1]
```diff
@@ -112,6 +123,64 @@ export async function addVisualizationPreference(
     await controller.runCell(cell);
 }
 
+export async function enterEarliestTime(
```
Comment (thesplunker) [inline out/notebooks/commands.ts:126]: nit: not a big deal at all, but these two functions seem to share logic. Totally ok to keep them this way if refactoring is pain.

File: out/notebooks/provider.ts [hunk 2]
```diff
@@ -64,6 +91,11 @@ export class CellResultCountStatusBarProvider implements vscode.NotebookCellStat
                             "arguments": [cell]
                         }
                     });
+                    // earliest and latest time pickers are only supported for SPL2 at the moment
```
Comment (thesplunker) [inline out/notebooks/provider.ts:94]: Why is this done in 2 places?

File: out/notebooks/commands.ts (hunk 1 above)
Comment (fantavlik) [reply out/notebooks/commands.ts:126]: I actually had the same thought and started to refactor it, but it ultimately only saved ~4 lines of code due to the multiple places where the two differ and ended up having a lot of indirection which made it harder to read so I reverted to this.

File: out/notebooks/provider.ts (hunk 2 above)
Comment (fantavlik) [reply out/notebooks/provider.ts:94]: One is for the state when the job hasn't been run yet and the other is for the state where there are outputs from the job. There are a different set of icons for each.

Before:
image

After (with job details):
image

File: out/notebooks/commands.ts (hunk 1 above)
Comment (thesplunker) [reply out/notebooks/commands.ts:126]: Honestly, then this is way better :) 4 lines ain't worth it

File: out/notebooks/spl2/installer.ts [hunk 3]
```diff
@@ -478,11 +487,11 @@ async function promptToDownloadLsp(alsoInstallJava: boolean): Promise<boolean> {
  */
 export async function getLatestSpl2Release(context: ExtensionContext, progressBar: StatusBarItem): Promise<void> {
     return new Promise(async (resolve, reject) => {
-        const lspArtifactPath = path.join(context.globalStorageUri.fsPath, 'lsp');
+        const lspArtifactPath = getLocalLspDir(context);
         // TODO: Remove this hardcoded version/update time and check for updates
-        let lspVersion: string = '2.0.362'; // context.globalState.get(stateKeyLatestLspVersion) || "";
+        let latestLspVersion: string = '2.0.362'; // context.globalState.get(stateKeyLatestLspVersion) || "";
         const lastUpdateMs: number = Date.now(); // context.globalState.get(stateKeyLastLspCheck) || 0;
```
Comment (thesplunker) [inline out/notebooks/spl2/installer.ts:493]: Should this be Date.now?

File: out/notebooks/spl2/serializer.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    async serializeNotebook(
+        data: vscode.NotebookData,
+        token: vscode.CancellationToken
+    ): Promise<Uint8Array> {
+        let contents: Spl2ModulesJson = <Spl2ModulesJson>{
+            modules: [],
+            app: "apps.search",
+        };
+
+        let indx = 1;
```
[the remark this answers is not in the document]
Comment (fantavlik) [reply out/notebooks/spl2/serializer.ts:67]: yeah I'm saving these as `module1`, `module2`, etc to be human readable.

**Review by thesplunker — APPROVED** [review]

File: out/notebooks/spl2/installer.ts (hunk 3 above)
Comment (fantavlik) [reply out/notebooks/spl2/installer.ts:493]: This is my way of turning the feature off for now until I have a chance to flesh out the issues I was seeing with retrieving the latest version from maven metdata, related to the `TODO` above

**Pull request merged**